src/bin/pred.rs

```diff
@@ -0,0 +1,200 @@
+use std::{fmt, hash::Hash};
+
+use egg::*;
+use rand::Rng;
+use rand_pcg::Pcg64;
+use ruler::*;
+
+define_language! {
+  pub enum Pred {
+    Lit(Constant),
+    Var(egg::Symbol),
+    "<" = Le([Id;2]),
+    "<=" = Leq([Id;2]),
+    ">" = Ge([Id;2]),
+    ">=" = Geq([Id;2]),
+    "==" = Eq([Id;2]),
+    "!=" = Neq([Id;2]),
+    "~" = Not(Id),
+    "&" = And([Id;2]),
+    "|" = Or([Id;2]),
+    "^" = Xor([Id;2]),
+  }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Constant {
+    Bool(bool),
+    Int(usize),
+}
+
+impl fmt::Display for Constant {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Constant::Bool(b) => b.fmt(f),
+            Constant::Int(i) => i.fmt(f),
+        }
+    }
+}
+
+impl std::str::FromStr for Constant {
+    type Err = ();
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Ok(i) = usize::from_str(s) {
+            Ok(Self::Int(i))
+        } else if let Ok(b) = bool::from_str(s) {
+            Ok(Self::Bool(b))
+        } else {
+            Err(())
+        }
+    }
+}
+
+impl Constant {
+    fn to_int(&self) -> Option<usize> {
+        match self {
+            Constant::Int(n) => Some(*n),
+            Constant::Bool(_) => None,
+        }
+    }
+
+    fn to_bool(&self) -> Option<bool> {
+        match self {
+            Constant::Bool(b) => Some(*b),
+            Constant::Int(_) => None,
+        }
+    }
+}
+
+impl SynthLanguage for Pred {
+    type Constant = Constant;
+
+    fn eval<'a, F>(&'a self, cvec_len: usize, mut v: F) -> CVec<Self>
+    where
+        F: FnMut(&'a Id) -> &'a CVec<Self>,
+    {
+        match self {
+            Pred::Lit(c) => vec![Some(c.clone()); cvec_len],
+            Pred::Le([x, y]) => {
+                map!(v, x, y => x.to_int().zip(y.to_int()).map(|(x,y)|Constant::Bool(x < y)))
```

Comment:
It can be a bit confusing if someone isn't pay attention to the types because the `$body` here resembles `map!`'s definition itself but I think this is the right way to do it anyway.

Reply:
Sorry not sure I follow? Is there a better way to do this?

Reply:
no I like it, this is good. Sorry my comment was kind of a stream of thought.